M365/MDO/ResendFailedMail.ps1

```diff
@@ -0,0 +1,322 @@
+        Write-Verbose "$_"
+        exit
+    }
+    if ($null -eq $graphConnection) {
+        Write-Host "Not connected to Graph" -ForegroundColor Red
+        Write-Host "Verify that you have Microsoft.Graph.Mail and Microsoft.Graph.Users.Actions modules installed and loaded." -ForegroundColor Yellow
+        Write-Host "You could use:" -ForegroundColor Yellow
+        Write-Host "`t`$ClientSecretCredential = Get-Credential -Credential ""[YOUR APP ID HERE]""" -ForegroundColor Yellow
+        Write-Host "`t# Enter client_secret in the password prompt." -ForegroundColor Yellow
+        Write-Host "`tConnect-MgGraph -TenantId ""[YOUR TENANT ID HERE]"" -ClientSecretCredential `$ClientSecretCredential -NoWelcome" -ForegroundColor Yellow
+        exit
+    } elseif ($graphConnection.count -eq 1) {
+        $expectedScopes = 'Mail.Read', 'Mail.Send'
+        if (Test-GraphContext -Scopes $graphConnection.Scopes -ExpectedScopes $expectedScopes) {
+            Write-Host "Connected to Graph"
+            Write-Host "Session details"
+            Write-Host "TenantID: $(($graphConnection).TenantId)"
+            Write-Host "AuthType: $(($graphConnection).AuthType)"
+        } else {
+            Write-Host "We cannot continue without Graph Powershell session without Expected Scopes." -ForegroundColor Red
+            Write-Host "Verify that you have Microsoft.Graph.Mail and Microsoft.Graph.Users.Actions modules installed and loaded." -ForegroundColor Yellow
+            Write-Host "You could use:" -ForegroundColor Yellow
+            Write-Host "`t`$ClientSecretCredential = Get-Credential -Credential ""[YOUR APP ID HERE]""" -ForegroundColor Yellow
+            Write-Host "`t# Enter client_secret in the password prompt." -ForegroundColor Yellow
+            Write-Host "`tConnect-MgGraph -TenantId ""[YOUR TENANT ID HERE]"" -ClientSecretCredential `$ClientSecretCredential -NoWelcome" -ForegroundColor Yellow
+            exit
+        }
+    } else {
+        Write-Host "You have more than one Graph sessions. Please use just one session." -ForegroundColor Red
+        exit
+    }
+    if (($graphConnection.TenantId) -ne ($exoConnection.TenantId) ) {
+        Write-Host "`nThe Tenant Id from Graph and EXO are different. Please use the same tenant." -ForegroundColor Red
+        exit
+    }
+}
+
+if ($PsCmdlet.ParameterSetName -eq 'ResendCopyFailedDays') {
+    $StartDate = (Get-Date).AddDays(-$Days)
+    Write-Verbose "StartDate: $StartDate"
+    $EndDate = Get-Date
+    Write-Verbose "EndDate: $EndDate"
+}
+
+$traceParams = @{
+    Status = "Failed"
+}
+
+if ($StartDate) { $traceParams["StartDate"] = $StartDate }
+if ($EndDate) { $traceParams["EndDate"] = $EndDate }
+if ($RecipientAddress ) { $traceParams["RecipientAddress"] = $RecipientAddress }
+if ($SenderAddress) { $traceParams["SenderAddress"] = $SenderAddress }
+if ($MessageId) { $traceParams["MessageId"] = $MessageId }
+
+$failedMessages = New-Object System.Collections.Generic.List[object]
```

Comment:
For all the `New-Object System.Collections.Generic.List[object]` you are creating a list, but then doing the action of `+=` to attempt to add to the list. However, this process in PowerShell will convert the variable to an `Array` which is a static list vs a dynamic one.

You actually only need to create a single list in the 3 locations that you have attempted to use this.